common/predictive-text/unit_tests/headless/edit-distance/distance-modeler.js

```diff
@@ -0,0 +1,225 @@
+
+      let secondEdge = queue.dequeue();
+      assert.equal(secondEdge.optimalInput[0].sample.insert, 'h');
+      assert.equal(secondEdge.calculation.lastMatchEntry.key, 'h');
+      assert.equal(secondEdge.currentCost, 0.75);
+
+      // After this, a 't' input without a matching char.
+      let nextEdge = queue.dequeue();
+      assert.equal(nextEdge.optimalInput[0].sample.insert, 't');
+      assert.notEqual(nextEdge.calculation.lastMatchEntry.key, 't');
+      assert.equal(nextEdge.currentCost, 1.25);
+    });
+
+    it('Small integration test:  "teh" => "ten", "the"', function() {
+      // The combinatorial effect here is a bit much to fully test.
+      let rootTraversal = testModel.traverseFromRoot();
+      assert.isNotEmpty(rootTraversal);
+
+      let rootNode = new correction.SearchNode(rootTraversal);
+      assert.equal(rootNode.calculation.getHeuristicFinalCost(), 0);
+
+      // VERY artificial distributions.
+      let synthDistribution1 = [
+        {sample: {insert: 't', deleteLeft: 0}, p: 1} // Transform, probability
+      ];
+
+      let synthDistribution2 = [
+        {sample: {insert: 'e', deleteLeft: 0}, p: 0.75}, // Transform, probability
+        {sample: {insert: 'h', deleteLeft: 0}, p: 0.25}
+      ];
+
+      let synthDistribution3 = [
+        {sample: {insert: 'h', deleteLeft: 0}, p: 0.75}, // Transform, probability
+        {sample: {insert: 'n', deleteLeft: 0}, p: 0.25}
+      ];
+
+      let layer1Edges = rootNode.buildSubstitutionEdges(synthDistribution1);
+      let layer1Queue = new models.PriorityQueue(correction.QUEUE_EDGE_COMPARATOR, layer1Edges);
+
+      let tEdge = layer1Queue.dequeue();
+      assertEdgeChars(tEdge, 't', 't');
+
+      let layer2Edges = new correction.SearchNode(tEdge).buildSubstitutionEdges(synthDistribution2);
+      let layer2Queue = new models.PriorityQueue(correction.QUEUE_EDGE_COMPARATOR, layer2Edges);
+
+      let eEdge = layer2Queue.dequeue();
+      assertEdgeChars(eEdge, 'e', 'e');
+
+      let hEdge = layer2Queue.dequeue();
+      assertEdgeChars(hEdge, 'h', 'h');
+
+      // Needed for a proper e <-> h transposition.
+      let ehEdge = findEdgeWithChars(layer2Edges, 'e', 'h');
+
+      assert.isOk(ehEdge);
+
+      // Final round:  we'll use three nodes and throw all of their results into the same priority queue.
+      let layer3eEdges  = new correction.SearchNode(eEdge).buildSubstitutionEdges(synthDistribution3);
+      let layer3hEdges  = new correction.SearchNode(hEdge).buildSubstitutionEdges(synthDistribution3);
+      let layer3ehEdges = new correction.SearchNode(ehEdge).buildSubstitutionEdges(synthDistribution3);
+      let layer3Queue = new models.PriorityQueue(correction.QUEUE_EDGE_COMPARATOR, layer3eEdges.concat(layer3hEdges).concat(layer3ehEdges));
+
+      // Find the first result with an actual word directly represented.
+      do {
+        edge = layer3Queue.dequeue();
+      } while(edge.calculation.lastMatchEntry.traversal.entries.length == 0);
```

Comment:
We want to filter out any partial words for this test.  (Otherwise, 'tel' for 'telephone', etc, may show up.)  I suppose I could have just built a small, pre-filtered fixture for use instead, though.

Reply:
Of course there will be.  I just don't want to handle them for _this_ test.  Please see #3534 and [its analogous test](https://github.com/keymanapp/keyman/pull/3534/files#diff-9b7c8cf4af112329598bd17d51826a82R233), which _does_ handle such cases.

Reply:
#3534 is in a much better spot to handle all the potential prefix & traversal variations and is far closer to the algorithm's eventual interface than the code seen here.